Support optional parameters at the end of the path

JulianBerger opened the pull request:
Originally wanted to add this feature to fastify: https://github.com/fastify/fastify/pull/2607 but we came to the conclusion that it fits better into here.

Support optional parameters (with a '?' suffix) if they are at the end of the path.

Example: 
**Input Route:** `/v1/user/:userId/admin/mail/:id?`
Can be accessed by:

1. `/v1/user/:userId/admin/mail/:id`
2. `/v1/user/:userId/admin/mail`

More information about the regex:
<img width="1592" alt="Screen Shot 2020-10-14 at 16 17 07" src="https://user-images.githubusercontent.com/3339615/96002228-2b75d780-0e39-11eb-8fa2-dfcc85b62ab4.png">

<img width="1599" alt="Screen Shot 2020-10-14 at 16 17 21" src="https://user-images.githubusercontent.com/3339615/96002207-287ae700-0e39-11eb-8d54-fbb42e1799eb.png">
So the trailing slash is still preserved with the second capture group.

delvedor commented on index.js:
```diff
@@ -70,6 +70,17 @@ Router.prototype.on = function on (method, path, opts, handler, store) {
   assert(typeof handler === 'function', 'Handler should be a function')
 
+  // path ends with optional parameter
+  const optionalParamRegex = /(\/:[^/]*?)\?(\/?)$/
```
Out of curiosity, why the need for a regex instead of `path.endsWith('?')`

JulianBerger replied:
I had `path.endsWith('?') || path.endsWith('?/')` before, but then it's not ensured to be a parameter.
That would also match for routes like `/foo?`, don't know if that is an issue..

delvedor replied:
Gotcha! 👍

delvedor reviewed (changes requested):
Can you add some additional test cases?
Eg:
- what happens if someone declares two following routes?
  - `/foo/:bar?`
  - `/foo`
- what happens if someone puts the question mark inside a route declaration?
  - `/foo/:bar?/baz`

It's useful to test also "bad behaviors", so we can be sure not to break anything in the future.

mcollina reviewed (approved):
lgtm

delvedor commented on test/optional-params.test.js:
```diff
@@ -0,0 +1,98 @@
+    }
+  })
+
+  findMyWay.on('GET', '/foo/:bar?/baz', (req, res, params) => {
```
Given that this case is not supported, I think we should throw an error here, what do you think?

JulianBerger replied:
Yes, it's throwing an Error now.

Eomm commented on index.js:
```diff
@@ -70,6 +70,17 @@ Router.prototype.on = function on (method, path, opts, handler, store) {
   assert(typeof handler === 'function', 'Handler should be a function')
 
+  // path ends with optional parameter
+  const optionalParamRegex = /(\/:[^/]*?)\?(\/?)$/
```
moving this regexp near to `FULL_PATH_REGEXP` would avoid to recreate it for every route/`on` call. It should optimize a little bit the GC execution + remove the duplicated declaration*

mcollina replied:
It would also avoid future bugs.

JulianBerger replied:
Good idea, declaring it globally now.

Eomm commented on index.js:
```diff
@@ -320,6 +331,17 @@ Router.prototype.off = function off (method, path) {
   assert(path[0] === '/' || path[0] === '*', 'The first character of a path should be `/` or `*`')
 
+  // path ends with optional parameter
+  const optionalParamRegex = /(\/:[^/]*?)\?(\/?)$/
```
the duplicated declaration*

JulianBerger reviewed (commented):
All resolved

mcollina reviewed (approved):
lgtm

JulianBerger commented:
@delvedor can you merge it please? :)

delvedor reviewed (changes requested):
Hello! Sorry for the delay :S
Everything looks good to me, can you add one last test to verify the behavior with `ignoreTrailingSlash: true`?

JulianBerger commented:
Added tests for ignoreTrailingSlash.

Eomm commented on test/optional-params.test.js:
```diff
@@ -0,0 +1,144 @@
+  })
+
+  findMyWay.lookup({ method: 'GET', url: '/a/foo-bar-baz', headers: {} }, null)
+  // findMyWay.lookup({ method: 'GET', url: '/a/foo', headers: {} }, null)
```
```suggestion
```

Eomm commented on test/optional-params.test.js:
```diff
@@ -0,0 +1,144 @@
+  const findMyWay = FindMyWay({
+    ignoreTrailingSlash: false,
+    defaultRoute: (req, res) => {
+      t.notMatch(req.url, '.*/$') // urls without trailing slash
```
```suggestion
      t.fail('Should not be defaultRoute')
```

delvedor commented on test/optional-params.test.js:
```diff
@@ -0,0 +1,144 @@
+    }
+  })
+
+  findMyWay.on('GET', '/test/hello/:optional?/', (req, res, params) => {
```
There is no need to add the trailing slash in the route definition, it might be passed by the client, which you are testing correctly.

```suggestion
  findMyWay.on('GET', '/test/hello/:optional?', (req, res, params) => {
```

delvedor commented on test/optional-params.test.js:
```diff
@@ -0,0 +1,144 @@
+    }
+  })
+
+  findMyWay.on('GET', '/test/hello/:optional?/', (req, res, params) => {
```
```suggestion
  findMyWay.on('GET', '/test/hello/:optional?', (req, res, params) => {
```

delvedor reviewed (changes requested):
Great work, one little nit, you should not register the route with the trailing slash, it should be used only in the "request".

JulianBerger commented:
> Great work, one little nit, you should not register the route with the trailing slash, it should be used only in the "request".

Ahh sure, fixed that! Also changed the ignoreTrailingSlash = false test behavior: `/test/hello/` will give us an empty string as the `:optional` parameter...

delvedor commented on test/optional-params.test.js:
```diff
@@ -0,0 +1,144 @@
+  })
+
+  findMyWay.on('GET', '/test/hello/:optional?', (req, res, params) => {
+    console.log('req', req, 'opt', params.optional)
```
```suggestion
```

delvedor commented on test/optional-params.test.js:
```diff
@@ -0,0 +1,144 @@
+
+  findMyWay.on('GET', '/test/hello/:optional?', (req, res, params) => {
+    console.log('req', req, 'opt', params.optional)
+    if (params.optional) {
```
This test looks brittle, can you write a more strict if check?

delvedor commented:
@JulianBerger Hello! Do you have time to finish this pr? :)

mcollina commented:
Closing for no activity

jcward commented:
FYI for anyone who lands here looking for optional trailing parameters in find-my-way -- it suited my needs to simply register the same handler twice:

```
    findMyWay.on('GET', '/list/:id', handle_list_items);
    findMyWay.on('GET', '/list', handle_list_items);
```
